File: src/lib/puzzle/solver.ts
import type { GridSize } from '../../types/puzzle';
import { isValidPlacement } from './validator';

export interface SolverResult {
  solved: boolean;
  solution: number[] | null;
  solutionCount?: number;
}

function findEmpty(cells: (number | null)[]): number {
  return cells.findIndex(cell => cell === null);
}

export function solve(
  cells: (number | null)[],
  size: GridSize
): SolverResult {
  const grid = [...cells];

  if (solveBacktrack(grid, size)) {
    return {
      solved: true,
      solution: grid as number[],
    };
  }

  return {
    solved: false,
    solution: null,
  };
}

function solveBacktrack(cells: (number | null)[], size: GridSize): boolean {
  const emptyIndex = findEmpty(cells);

  if (emptyIndex === -1) {
    return true; // Puzzle solved
  }

  // Try values 1 to size
  for (let value = 1; value <= size; value++) {
    if (isValidPlacement(cells, emptyIndex, value, size)) {
      cells[emptyIndex] = value;

      if (solveBacktrack(cells, size)) {
        return true;
      }

      cells[emptyIndex] = null;
    }
  }

  return false;
}

export function countSolutions(
  cells: (number | null)[],
  size: GridSize,
  maxCount: number = 2
): number {
  const grid = [...cells];
  let count = 0;

  function backtrack(): boolean {
    const emptyIndex = findEmpty(grid);

    if (emptyIndex === -1) {
      count++;
      return count >= maxCount;
    }

    for (let value = 1; value <= size; value++) {
      if (isValidPlacement(grid, emptyIndex, value, size)) {
        grid[emptyIndex] = value;

        if (backtrack()) {
          grid[emptyIndex] = null;
          return true;
        }

        grid[emptyIndex] = null;
      }
    }

    return false;
  }

  backtrack();
  return count;
}

export function hasUniqueSolution(
  cells: (number | null)[],
  size: GridSize
): boolean {
  return countSolutions(cells, size, 2) === 1;
}

export function solveWithRandomization(
  cells: (number | null)[],
  size: GridSize
): SolverResult {
  const grid = [...cells];

  if (solveBacktrackRandomized(grid, size)) {
    return {
      solved: true,
      solution: grid as number[],
    };
  }

  return {
    solved: false,
    solution: null,
  };
}

function solveBacktrackRandomized(
  cells: (number | null)[],
  size: GridSize
): boolean {
  const emptyIndex = findEmpty(cells);

  if (emptyIndex === -1) {
    return true;
  }

  // Shuffle values for randomization
  const values = shuffle(Array.from({ length: size }, (_, i) => i + 1));

  for (const value of values) {
    if (isValidPlacement(cells, emptyIndex, value, size)) {
      cells[emptyIndex] = value;

      if (solveBacktrackRandomized(cells, size)) {
        return true;
      }

      cells[emptyIndex] = null;
    }
  }

  return false;
}

function shuffle<T>(array: T[]): T[] {
  const result = [...array];
  for (let i = result.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [result[i], result[j]] = [result[j], result[i]];
  }
  return result;
}

File: love/src/puzzle/solver.lua
-- Backtracking solver for 2D sudoku.
-- Operates on 1-indexed flat arrays; nil = empty cell.

local V = require "src.puzzle.validator"

local Solver = {}

local function shuffle(t)
  for i = #t, 2, -1 do
    local j = math.random(i)
    t[i], t[j] = t[j], t[i]
  end
  return t
end

local function find_empty(cells, n)
  for i = 1, n * n do
    if cells[i] == nil then return i end
  end
  return nil
end

-- Deterministic solve. Returns true and fills cells in-place on success.
local function backtrack(cells, n)
  local idx = find_empty(cells, n)
  if not idx then return true end

  for v = 1, n do
    if V.is_valid_placement(cells, idx, v, n) then
      cells[idx] = v
      if backtrack(cells, n) then return true end
      cells[idx] = nil
    end
  end
  return false
end

-- Randomised solve (for puzzle generation). Returns true and fills cells in-place.
local function backtrack_random(cells, n)
  local idx = find_empty(cells, n)
  if not idx then return true end

  local values = {}
  for v = 1, n do values[v] = v end
  shuffle(values)

  for _, v in ipairs(values) do
    if V.is_valid_placement(cells, idx, v, n) then
      cells[idx] = v
      if backtrack_random(cells, n) then return true end
      cells[idx] = nil
    end
  end
  return false
end

-- Count solutions up to `limit` (default 2 — enough to detect uniqueness).
local function count_solutions(cells, n, limit)
  limit = limit or 2
  local count = 0

  local function bt()
    local idx = find_empty(cells, n)
    if not idx then
      count = count + 1
      return count >= limit
    end
    for v = 1, n do
      if V.is_valid_placement(cells, idx, v, n) then
        cells[idx] = v
        if bt() then
          cells[idx] = nil
          return true
        end
        cells[idx] = nil
      end
    end
    return false
  end

  bt()
  return count
end

-- Public API ──────────────────────────────────────────────────────────────────

-- Solve deterministically. Returns solved copy or nil.
function Solver.solve(cells, n)
  local grid = {}
  for i = 1, n * n do grid[i] = cells[i] end
  if backtrack(grid, n) then return grid end
  return nil
end

-- Solve with randomised value order (for generation). Returns filled copy or nil.
function Solver.solve_random(cells, n)
  local grid = {}
  for i = 1, n * n do grid[i] = cells[i] end
  if backtrack_random(grid, n) then return grid end
  return nil
end

-- Returns true if the puzzle has exactly one solution.
function Solver.has_unique_solution(cells, n)
  local grid = {}
  for i = 1, n * n do grid[i] = cells[i] end
  return count_solutions(grid, n, 2) == 1
end

return Solver
